landlab/components/species_evolution/record.py

```diff
@@ -0,0 +1,190 @@
+    ``set_value``. Variable values not set at a time have a value of `nan`.
+    """
+    def __init__(self, initial_time=0):
+        """Instantiate Record.
+
+        Parameters
+        ----------
+        initial_time : float or int
+            The initial time in the record.
+        """
+        self._dict = OrderedDict([('time', [initial_time])])
+
+    @property
+    def times(self):
+        """The times stored in the record."""
+        return self._dict['time']
+
+    @property
+    def prior_time(self):
+        """The penultimate time in the record.
+
+        nan is returned if the record stores only one time step.
+        """
+        if len(self.times) < 2:
+            return np.nan
+        else:
+            return sorted(self.times)[-2]
+
+    @property
+    def earliest_time(self):
+        """The earliest time in the record."""
+        return min(self.times)
+
+    @property
+    def latest_time(self):
+        """The latest time in the record."""
+        return max(self.times)
+
+    @property
+    def count_of_time_steps(self):
+        """The count of record time steps."""
+        return len(self._dict['time'])
+
+    @property
+    def variables(self):
+        """The variables in the record."""
+        variables = list(self._dict.keys())
+        variables.remove('time')
+        return variables
+
+    @property
+    def data_frame(self):
+        """A Pandas DataFrame of the record."""
+        return DataFrame(self._dict)
+
+    def __len__(self):
+        """The count of record time steps."""
+        return self.count_of_time_steps
+
+    def advance_time(self, dt):
+        """Advance the time in the record by a time step duration.
+
+        This method creates a new record entry. The new entry includes the time
+        and all variables have a value of nan for this time. The time of the
+        entry is the sum of the latest time of the record and the time step
+        duration, ``dt``.
+
+        Parameters
+        ----------
+        dt : float or int
+            The time step duration to advance time in the record.
+        """
+        self._dict['time'].append(self.latest_time + dt)
+
+        for var in self.variables:
+            self._dict[var].append(np.nan)
+
+    def get_value(self, var_name, time=np.nan):
+        """Get the value of a variable.
+
+        Parameters
+        ----------
+        var_name : string
+            The name of the variable to get.
+        time : float or int, optional
+            The time of the variable to get. The latest time in the record is
+            the default.
+
+        Returns
+        -------
+        T
+            The value of `var_name` that has the type, T. `nan` is returned if
+            `var_name` does not exist in the record.
+        """
+        if var_name not in self.variables:
+            return np.nan
+
+        time = self._check_time(time)
+
+        idx = self._get_time_index(time)
+        return self._dict[var_name][idx]
```

Comment:
This, along with `set_value` and `increment_value` require that one knows the type of a specific var, so as not to set it incorrectly.  Is a `get_type` overkill? Maybe...

Reply:
...maybe or maybe not. The record is not exposed to users. Its just handy to use with Taxon types. So only those gaining deeper familiarity with SpeciesEvolver will use Record and will by then know the types. I've made note to consider this down the line.

Reply:
That helps. Because I'm just reading code who sees what isn't obvious to me yet.